File: Medium/182. Delete Digits/Solution.cpp
class Solution {
public:
    /**
     * @param A: A positive integer which has N digits, A is a string
     * @param k: Remove k digits
     * @return: A string
     */
    string DeleteDigits(string &A, int k) {
        // write your code here
        string res;
        int keep = A.size() - k;
        for (int i=0; i<A.size(); i++) {
            while (res.size()>0 && res.back()>A[i] && k>0) {
                res.pop_back();
                k--;
            }
            res.push_back(A[i]);
        }
        res.erase(keep, string::npos);

        // trim leading zeros
        int s = 0;
        while (s<(int)res.size()-1 && res[s]=='0')  s++;
        res.erase(0, s);

        return res=="" ? "0" : res;
    }
};

File: Medium/182. Delete Digits/Solution.py
class Solution:
    """
    @param A: A positive integer which has N digits, A is a string
    @param k: Remove k digits
    @return: A string
    """
    def DeleteDigits(self, A, k):
        # write your code here
        A = list(A)
        while k > 0:
            f = True
            for i in xrange(len(A)-1):
                if A[i] > A[i+1]:
                    del A[i]
                    f = False
                    break
            if f and len(A)>1:
                A.pop()
            k -= 1
        while len(A) > 1 and A[0]=='0':
            del A[0]
        return ''.join(A)
